### DISCUSSION TITLE
My thoughts on this, bugs, and suggestions.

### DISCUSSION THREAD

**Discussion by XanCraft21** [board707/w80x_arduino#19] [General]
Hello, i would like to share my thoughts on this software package. I am so glad you made this, i was struggling to find an Arduino package for the W806 controllers, the only one i could find was abandoned after only 3 basic functions were implemented. But when i found this one i was immediately shocked, this one may be incomplete but it has all the most common functions completed, with only 3 extra functions to go.

There are some bugs i need to discuss, i found about 2. The first bug is the set_loop function in SPI.cpp doesn’t seem to exist, so i had to manually comment it out. Looking at the files on this page it appears to be already commented out, but i did not receive that when i downloaded it.

The second bug is with the compiler tools, specifically the way functions are handled. It appears that this software runs in c++, but a bug breaks the automatic function wrapping within Arduino, so the functions need to be made as if this were using pure c.

There are a few things I disagree with in the design choices of the software, like the SPI and Wire libraries. The ones here are programmed in a nonstandard way, which causes compatibility issues with most libraries like Adafruit display and graphics libraries. Surprisingly u8g2 still works fine. I was wondering if it were possible for you to reprogram the SPI and Wire libraries to make them like the other ones in other boards packages.

I’m really glad i found this package, i thought all the winner micro chips were all junk, but now i can get a small amount of use out of them. Thank you for making this package, you finally made useless electronics become useful.

**Comment by board707** [comment]
Hi, @XanCraft21 !
First of all, thank you very much for your feedback and reporting problems. This information is very important to us, as interest in this package is still modest and we are in need of bug reports.
As for the `set_loop` function, this bug has already been fixed in the development branch, but is not yet included in the master release of the project.
The next problem, the C and C++ function mix - it is an issue for us too, but that's how Winner Micro's SDK was originally written, we can't do much with it.

>  SPI and Wire libraries. The ones here are programmed in a nonstandard way, which causes compatibility issues with most libraries like Adafruit display and graphics libraries. 

Could you please explain your opinion in more detail? What does non-standard mean and how would you suggest changing them? If possible, please provide an example of a library that is experiencing compatibility issues due to the non-standard design of our package.

Thank you again. If you have more bugs to describe  - please feel free to let us know.
We will be grateful if you rate our work on github.

**Reply by XanCraft21** [reply]
You are very welcome for the feedback.

what i mean by non-standard design of the SPI and Wire libraries is something to do with the class name. I do not remember what needs to be changed in the Wire library, but for the SPI library it is the class name. If i look at the files i see the names are base_spi, HardSPI, and SoftSPI. If it would be possible to change it to just be SPIClass so other libraries that use SPI will work with this.

The libraries that don’t work i’ve only tested 2 so far, Adafruit_GFX and their display driver libraries Adafruit_SSD1306 and Adafruit_ST7735/89. And the second one goes by 2 names, Arduino_GFX and GFX_Library_For_Arduino made by moononournation.

There are 2 problems with the Adafruit libraries, the first one is on the library side where is checks for undefined boards before including a delay utility, which doesn’t take into consideration for the W80x boards. I can comment it out just fine, but the class names for the SPI library do not match.

For Arduino_GFX i think it’s the same problem with the class names, but i also think the library checks for specific board types before including core drivers.

I only tested the examples within the libraries but when i can i will get the code for you if you need it.

But the u8g2 library works fine, but i think that library handles the interfaces a bit differently.

What i think is for the SPI library the class name just needs to be change to be called “SPIClass” as it is in other board packages. Thank you for taking the time to look at this. I hope i’m not giving you too much hastle.

**Reply by board707** [reply]
Thanks for sharing your thoughts. I agree that SPI should be compatible with SPIClass and I'll see how that can be done.
As for the problems with running Adafruit libraries - it's not just about SPI. I have found some more bugs in our package which will hopefully be fixed in the next release.
If you have any other comments on our package - please feel free to report in discussions or in the issues.
WBR, 
Dmitry

**Reply by XanCraft21** [reply]
Thank you, i really hope to see this repo become very popular and be finished. It’s more finished than the one from Hi-LinkDuino and nulllabs.

Another suggestion that may or may not be possible, can you add support for the W600 chipset? I found one on AliExpress in the form of an ESP-01 module with possible wifi functionality by the seller WAVGAT. If i read the specs in the description correctly the W600 has an ARM Cortex-m3 core and i don’t know how easy it will be to make that work.

I personally do not feel like trying new software to make the W600 chipset work, as i personally prefer using Arduino for ease of access and simplicity reasons.

**Reply by board707** [reply]
There is an Arduino package for W600 controllers on githab. But I haven't tried it

https://github.com/w600/arduino

**Reply by XanCraft21** [reply]
Thank you, but unfortunately it appears that the project has been abandoned 4 years ago and the boards manager link no longer works.

**Reply by AnatolSher** [reply]
See https://raw.githubusercontent.com/salmanfarisvp/snippet/master/package_wmcom_index.json 
![image](https://github.com/board707/w80x_arduino/assets/121094544/941b22d0-4684-4228-9b41-1392b7fdcfa7)
Download it, then update from https://github.com/w600/arduino if you need releise candidat

Python for w600
[w600_Python.zip](https://github.com/board707/w80x_arduino/files/11992240/w600_Python.zip)
![image](https://github.com/board707/w80x_arduino/assets/121094544/e56eba24-85d6-44cc-8812-54d0474da14c)
I experimented with w600-pico

**Reply by XanCraft21** [reply]
Thank you, that is very helpful, but i’d rather wait a while and see if something new comes up in development. If i’m reading this right i create my sketch code and i use a separate python script to upload the code? I personally prefer the scripts being within the Arduino boards package, but i will look at this more.

Or is the python script for making the code in python? I seem to be misunderstanding what you said, I apologize for that.

**Reply by AnatolSher** [reply]
Python is different, Arduino is different. If you want python see https://www.wemos.cc/en/latest/w600/w600_pico.html.
https://www.wemos.cc/en/latest/tutorials/w600/get_started_with_micropython_w600.html
Download the runtime (one of the files in the w600_Python.zip archive) and write the code in Thonny IDE

For arduino new assemblies are not expected. Latest version is 0.2.6

**Reply by XanCraft21** [reply]
Thank you, i will choose to use Arduino since i have never worked with micropython before.

I downloaded the package but i can’t compile because one of the tools is bugged. I get a popup window saying “Microsoft Visual C++ Runtime Library Debug Assertion Failed!” I get 3 different popups for line 99 in the debug tool but they’re all the same, Exception (stream != null). Do you know how to fix this? Do i need to open an issue in the repo you linked me to. Thank you for your help.

This is off topic for this repo and we may need to move somewhere else.

**Reply by board707** [reply]
@erikbower65 
It seems to me that your message is off-topic for this discussion.

**Reply by XanCraft21** [reply]
I think he accidentally sent that in the wrong discussion wall, no worries.

Anyways I really like the extended optimization settings added in the latest update of this repo. I no longer need to open the boards txt file to edit that value.

Also board707 do you have plans to make a W600 boards package for Arduino? The only one that exists has broken links and is entirely bugged and cannot finish even the verify compile option.

**Reply by board707** [reply]
> do you have plans to make a W600 boards package for Arduino? 

Honesty speaking, no. At least at the nearest future. I haven't w600 boards for testing. Also I don't see it in the market now, so I think the w600 package have lost most of actuality now.

**Reply by AnatolSher** [reply]
https://aliexpress.ru/item/32978979389.html?sku_id=66800812387&spm=a2g2w.productlist.search_results.2.504a6a8dvqVfcZ
У этой модификации чипа w600 - 2 M флеша на борту, а периферия как у w80x

**Reply by XanCraft21** [reply]
Ahh, that sucks. But there are very few boards with w600 out there, and the official support is slightly lacking. There may be an SDK but no good tools for uploading.

AnatolSher posted a link to a w600 product on AliExpress. I have also found one that looks like an esp-01, originally from the esp8266. (Note esp8266 is way better, and also try the esp32-c3 version by lilygo.)

https://a.aliexpress.com/_mOl2n1s
